Added IndexSpanAlias and IndexServiceAlias for explicit aliases

SomilJain0112 opened the pull request:
## Which problem is this PR solving?
Resolves #7223

## Description of the changes
 a. Add IndexSpanAlias and IndexServiceAlias config fields
 b. Use explicit aliases when provided instead of prefix pattern
   
## How was this change tested?
Through local testing 

## Checklist
- [✅ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ✅] I have signed all commits
- [ ✅] I have added unit tests for the new functionality
- [✅ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `npm run lint` and `npm run test`

SomilJain0112 commented:
Hii @yurishkuro ,
I have resolved all the comments made by you, please review the changes. 
Thanks!

SomilJain0112 commented:
Hii @yurishkuro , I have resolved comments. Please review here!

yurishkuro commented on internal/storage/v1/elasticsearch/spanstore/reader.go:
```diff
@@ -114,47 +114,71 @@ type SpanReader struct {
+		serviceIndexPrefix = p.ServiceIndexOverride
+	}
+
+	useAliasMode := p.UseReadWriteAliases || hasSpanOverride || hasServiceOverride
```
```suggestion
	useAliasMode := p.UseReadWriteAliases
```

yurishkuro commented on internal/storage/v1/elasticsearch/spanstore/reader.go:
```diff
@@ -114,47 +114,71 @@ type SpanReader struct {
 	}
 
+	var timeRangeFn TimeRangeIndexFn
+	if hasSpanOverride || hasServiceOverride {
```
spaghetti. Keep all index logic in TimeRangeIndicesFn

yurishkuro commented on internal/storage/v1/elasticsearch/spanstore/writer.go:
```diff
@@ -103,12 +105,37 @@ type spanAndServiceIndexFn func(spanTime time.Time) (string, string)
+	hasSpanOverride := p.SpanIndexOverride != ""
+	if hasSpanOverride {
+		spanIndexPrefix = p.SpanIndexOverride
+	}
```
why is this needed? There is a very simple function in L133 `if p.UseReadWriteAliases {` which covers the overrides condition and where overrides can be applied instead of composing index name from pieces

SomilJain0112 commented:
Hii @yurishkuro , 
Addressed your above comments and made changes accordingly in the code. Kindly review it. 
Thanks!

yurishkuro commented on internal/storage/elasticsearch/config/config.go:
```diff
@@ -126,6 +126,14 @@ type Configuration struct {
+	// SpanIndexOverride allows full control over the name of the index alias.
+	// Overrides any index prefix / suffix settings.
+	// Can only be used with UseReadWriteAliases=true.
+	SpanIndexOverride string `mapstructure:"span_index_override"`
```
> When set, it replaces the prefix but read/write suffixes are still appended.

doesn't that go against the requirement expressed by the user?

SomilJain0112 commented:
Hii @yurishkuro @mahadzaryab1 
Please review this PR. 
Thanks!

yurishkuro reviewed (commented):
sorry for delay. Looks good, with a couple of minor comments

SomilJain0112 commented:
Hii @yurishkuro 
Removed redundant checks as it was not needed, removed irrelevant comment as well and removed an unnecessary testcase as well. Resolved all of your latest comments. Please have a look again. 
Thanks!
